ArgumentCoderTests should provide facilities for encoder-generic typed testing suites

zdobersek opened the pull request:
#### 99c0f8b4cfbdecaae4054afb4973c8addf33c659
<pre>
ArgumentCoderTests should provide facilities for encoder-generic typed testing suites
<a href="https://bugs.webkit.org/show_bug.cgi?id=249868">https://bugs.webkit.org/show_bug.cgi?id=249868</a>

Reviewed by Kimmo Kinnunen.

In ArgumentCoderTests unit testing file, ArgumentCoderEncoderDecoderTest is
turned into a template of which two specializations are provided, one for
IPC::Encoder and the other for IPC::StreamConnectionEncoder. This enables test
fixtures that can be used in test cases that cover functionality of both
encoders.

The existing set of ArgumentCoderDecodingMoveCounterTest tests is converted into
one such suite, now covering both types of encoders.

* Tools/TestWebKitAPI/Tests/IPC/ArgumentCoderTests.cpp:
(TestWebKitAPI::EncoderDecoderTest::name):
(TestWebKitAPI::EncoderTypeNames::GetName):
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest&lt;IPC::Encoder&gt;::encoder const):
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest&lt;IPC::Encoder&gt;::headerSize const):
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest&lt;IPC::Encoder&gt;::encoderSize const):
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest&lt;IPC::Encoder&gt;::createDecoder const):
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest&lt;IPC::StreamConnectionEncoder&gt;::encoder const):
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest&lt;IPC::StreamConnectionEncoder&gt;::headerSize const):
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest&lt;IPC::StreamConnectionEncoder&gt;::encoderSize const):
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest&lt;IPC::StreamConnectionEncoder&gt;::createDecoder const):
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest&lt;IPC::StreamConnectionEncoder&gt;::Impl::Impl):
(TestWebKitAPI::EncodingCounter::encode const):
(TestWebKitAPI::EncodingCounter::encode):
(TestWebKitAPI::DecodingMoveCounter::encode):
(TestWebKitAPI::DecodingMoveCounter::decode):
(TestWebKitAPI::TYPED_TEST_P):
(TestWebKitAPI::DecodingMoveCounterTest::name): Deleted.
(TestWebKitAPI::ArgumentCoderDecodingMoveCounterTest::encoder const): Deleted.
(TestWebKitAPI::ArgumentCoderDecodingMoveCounterTest::createDecoder const): Deleted.
(TestWebKitAPI::TEST_F): Deleted.

Canonical link: <a href="https://commits.webkit.org/258668@main">https://commits.webkit.org/258668@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/539865ccbf5db81386124fb49642a0fc1c3db9c0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102626 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11756 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35661 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111893 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172116 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106594 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12760 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/2662 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94901 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109596 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108402 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9788 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/93005 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37448 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91642 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24514 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79192 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5212 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25933 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5372 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2383 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11393 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45422 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5952 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7102 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Tools/TestWebKitAPI/Tests/IPC/ArgumentCoderTests.cpp:
```diff
@@ -28,10 +28,103 @@
+    }
+
+    IPC::StreamConnectionEncoder& encoder() const { return m_impl->encoder; }
+    size_t headerSize() const { return 16; }
```
I don't think this is used now? Now that the header is actually faked above, it's a bit contrived to have the function. Unless there's a grand plan for the function in future, maybe you can omit it..

zdobersek replied:
Can be removed.

zdobersek replied:
It's actually needed in follow-up Span tests.

kkinnunen-apple commented on Tools/TestWebKitAPI/Tests/IPC/ArgumentCoderTests.cpp:
```diff
@@ -28,10 +28,103 @@
+        WTF_MAKE_STRUCT_FAST_ALLOCATED;
+
+        Impl()
+            : buffer(1024, uint8_t(0))
```
stylistically this could perhaps be a static cast or maybe the auto conversion works? or then ` { } `.

zdobersek replied:
Needs the static cast, otherwise it collides with other constructors because of the `0` being ambiguous.

kkinnunen-apple commented on Tools/TestWebKitAPI/Tests/IPC/ArgumentCoderTests.cpp:
```diff
@@ -28,10 +28,103 @@
+        // the encoded size up to 16 bytes.
+        m_impl = makeUnique<Impl>();
+        ASSERT_EQ(m_impl->encoder.size(), 2u);
+        m_impl->encoder << uint64_t(0);
```
stylistically this probably should still be static cast ..

Also, the intention is that everything works with the way the encoder is setup by default. E.g. the extra 64-bit value could be hiding some errors.

I think better strategy would be to not put the filler in. Rather write the tests and expect them to work.

If some test would need the encoder / size, you could adjust the size in the encoder size accessor. Like, `encodedDataSize()` or something. 

By quick finds, I couldn't determine if any tests in this patch would fail unintentionally due to the filler missing?

Of course, using the "encoded data size" is tricky, because the size of the alignment skip will depend on what is the first element being encoded / expected to be decoded. If you don't see a solution immediately, maybe we can ignore this suggestion and use the hardcoded filler.

Note: the uint64_t doesn't entirely circumvent the issue, as I belive max_align is larger than 16 bytes. E.g. there exist a type that induces a skip, so you'd encode the type and then assert a value larger than the type, which is surprising. However, the primary purpose of the filler, making StreamConnectionEncoder and Encoder invocations use the same value, would still be fulfilled.

zdobersek replied:
I can remove the filler. Subsequent tests can be rewritten to effectively manually round up and adjust the expected encoder size, which should be fine.

So the filler was used to accommodate the difference in header size between the two encoders, testing from a common encoder size. Funnily enough the `IPC::Encoder` header size was then increased a few days ago due to adoption of 128-bit identifiers.

zdobersek committed:
ArgumentCoderTests should provide facilities for encoder-generic typed testing suites
https://bugs.webkit.org/show_bug.cgi?id=249868

Reviewed by Kimmo Kinnunen.

In ArgumentCoderTests unit testing file, ArgumentCoderEncoderDecoderTest is
turned into a template of which two specializations are provided, one for
IPC::Encoder and the other for IPC::StreamConnectionEncoder. This enables test
fixtures that can be used in test cases that cover functionality of both
encoders.

The existing set of ArgumentCoderDecodingMoveCounterTest tests is converted into
one such suite, now covering both types of encoders.

* Tools/TestWebKitAPI/Tests/IPC/ArgumentCoderTests.cpp:
(TestWebKitAPI::EncoderDecoderTest::name):
(TestWebKitAPI::EncoderTypeNames::GetName):
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest<IPC::Encoder>::encoder const):
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest<IPC::Encoder>::headerSize const):
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest<IPC::Encoder>::encoderSize const):
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest<IPC::Encoder>::createDecoder const):
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest<IPC::StreamConnectionEncoder>::encoder const):
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest<IPC::StreamConnectionEncoder>::headerSize const):
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest<IPC::StreamConnectionEncoder>::encoderSize const):
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest<IPC::StreamConnectionEncoder>::createDecoder const):
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest<IPC::StreamConnectionEncoder>::Impl::Impl):
(TestWebKitAPI::EncodingCounter::encode const):
(TestWebKitAPI::EncodingCounter::encode):
(TestWebKitAPI::DecodingMoveCounter::encode):
(TestWebKitAPI::DecodingMoveCounter::decode):
(TestWebKitAPI::TYPED_TEST_P):
(TestWebKitAPI::DecodingMoveCounterTest::name): Deleted.
(TestWebKitAPI::ArgumentCoderDecodingMoveCounterTest::encoder const): Deleted.
(TestWebKitAPI::ArgumentCoderDecodingMoveCounterTest::createDecoder const): Deleted.
(TestWebKitAPI::TEST_F): Deleted.

Canonical link: https://commits.webkit.org/258668@main
